### DISCUSSION TITLE
Enforce encrypted zip for unstreaming when password parameter is supplied

### DISCUSSION THREAD

**Discussion by HugoMichelCognex** [uktrade/stream-unzip#96] [Ideas]
Hey,

I just found out using your library (which is really nice btw, thank you very much) that even if you specify a password in the `stream_unzip` method, you can still send zip that are not encrypted and they are still being read.

While this can be a nice feature, it could also be a problem if you want to control the input zip that is being passed. In our case, we want to be sure the input zip has been produced by us and is therefore encrypted with our password.

Could it be possible to add either an option or a flag to raise an error if you try to read a non-encrypted archive when you supply a password? I have not fully read your code, but this should (?) be possible I guess.

Thanks 🙏

**Comment by michalc** [comment]
It should be possible yes. And in fact I wonder if this should be the default behaviour, and maybe even _not_ turn-offable... it'll be technically a backwards-incompatible change, but I suspect that's fine.

For a bit of background, what sort of encryption do your zip files use? Old-style ZipCrypto or AES? (And if AES, what key length encryption key if you know?)

**Reply by michalc** [reply]
Although maybe it _should_ be turn-offable...  Could add a parameter allowed_encryption_mechnisms that by default only allows the strongest encrpytion if a password is supplied:

```python
allowed_encryption_mechnisms=(
  AES_256,
)
```
But could be changed to allow others:
```python
allowed_encryption_mechnisms=(
  NONE,
  ZIP_CRYPTO,
  AWS_128,
  AES_192,
  AES_256,
) 
```

**Reply by HugoMichelCognex** [reply]
I only tried with ZipCrypto for now.

**Reply by michalc** [reply]
Have got a WIP version at https://github.com/uktrade/stream-unzip/pull/97

**Reply by michalc** [reply]
The WIP version at https://github.com/uktrade/stream-unzip/pull/97 now has a bit of documentation explaining how to tighten up the allowed encryption mechanism

**Reply by michalc** [reply]
Merged in and released in v0.0.93. So in the case discussed here, to only allow ZipCrypto, but _not_ allow anything else, including non-encrypted files, you would do something like this:

```python
from stream_unzip import stream_unzip, ZIP_CRYPTO

for file_name, file_size, unzipped_chunks in stream_unzip(
    zipped_chunks(),
    password=b'my-password', allowed_encryption_mechanisms=(ZIP_CRYPTO,),
):
```
And it does not allow non-encrypted files because `NO_ENCRYPTION` was not passed in the allowed_encryption_mechanisms parameter.

**Reply by HugoMichelCognex** [reply]
That is perfect, thank you very much!!

Your responsiveness is amazing 👍
